add float semantics RFC (#3514)
rust-lang/rfcs: text/0000-float-semantics.md

> - Feature Name: `float_semantics`

Reviewer: Maybe it should be `float_nan_semantics` instead? There is more to float semantics than just NaNs, like for example there could be RFCs about changing Rust from IEEE 754-2008 to IEE 754-2019 (see the maximum vs maxnum thread in https://github.com/rust-lang/rust/issues/83984 for some discussion on this for example).

Otherwise this reminds me of #1857 which if you read the RFC only stabilized a little part of Rust's drop order logic, but got me confused at the start, making me think that say `&&` chain drop order was stable (it was not and I changed it).

Author: We also say that outside NaNs we exactly match IEEE. In particular, we will use exactly the precision declared in the source. In particular, this RFC explicitly says we do *not* do what https://github.com/rust-lang/rfcs/pull/2686 proposed.

So I think the name is adequate.

- Start Date: 2023-10-14
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Rust's floating point operations follow IEEE 754-2008 -- with some caveats around operations producing NaNs: IEEE makes almost no guarantees about the sign and payload bits of the NaN; however, actual hardware does not pick those bits completely arbitrarily, and Rust will expose some of those hardware-provided guarantees to programmers.
On the flip side, NaN generation is non-deterministic: running the same operation on the same inputs several times can produce different results.
And there is a caveat: while IEEE specifies that float operations can never output a signaling NaN, Rust float operations *can* produce signaling NaNs, *but only if* an input is signaling.

> That means the only way to ever see a signaling NaN in a program is to create one with `from_bits` (or equivalent unsafe operations).

Reviewer 2: From a reader's perspective, this part doesn't differentiate 'current behavior' and 'proposed behavior' well?

Author: The RFC describes the proposed spec.

Current situation is that we have to spec so there's not really anything to compare with.
Current behavior as implemented by rustc already satisfies the proposed spec.

That effort should only be invested if there is a significant payoff.
The RFC author considers the downsides of unspecified NaN bit patterns being observable in const-evaluation to be minimal, and hence the payoff of this alternative to be low.

# Prior art
[prior-art]: #prior-art

C and C++ generally avoid saying anything about signaling NaNs, and also don't talk about how NaNs are chosen. I can only assume this means they are chosen non-deterministically; that is certainly the interpretation chosen by GCC and LLVM.

GCC [says](https://gcc.gnu.org/wiki/FloatingPointMath) "Without any explicit options, GCC assumes round to nearest or even and does not care about signalling NaNs". It is unclear whether "does not care" also means "guarantees to never produce by itself", i.e. whether `0.0 / 0.0` is ever allowed to evaluate to a signaling NaN or not. If it *is* allowed to evaluate to a signaling NaN, that is probably a violation of the C standard, which guarantees that `pow(1, 0.0/0.0)` returns `1` -- but in practice, `pow(1, sNaN)` returns a NaN.

> LLVM [recently adopted](https://github.com/llvm/llvm-project/pull/66579) new NaN rules that this RFC copies exactly into Rust.

Reviewer 3: since MSVC is one of the tier-one targets, does it have anything to say about these things?

Author: The MSVC ABI is one of our tier 1 targets, but that's orthogonal to what the MSVC compiler does for float operations in C functions it compiles.

I don't know what MSVC does wrt float guarantees since I'm completely disconnected from the Windows ecosystem; if someone has that information and can fill it in, please let me know. :)

Reviewer 4: You'd probably want to see [the docs](https://learn.microsoft.com/en-us/cpp/build/reference/fp-specify-floating-point-behavior?view=msvc-170) for guarantees. `/fp:precise` is the default.

Author: Thanks for the link! It raises about as many questions as it answers. :)

- They talk about "source precision" and "machine precision". So the result of `a + b + c` might be different from what IEEE says since intermediate results might be stored at a higher (or lower?) precision than that of the source type.
- It says that rounding to source precision is done at function calls and returns. So their inliner must preserve these function boundaries then to ensure the value doesn't stay at machine precision? Also, does "rounding to source precision" imply "NaN bits can change"?
- They say "The compiler doesn't perform algebraic transformations on floating-point expressions, such as reassociation or distribution, unless it can guarantee the transformation produces a bitwise identical result". Does this mean they do not even apply commutativity? Commutativity does *not* produce a bitwise identical result on actual hardware if you take NaNs into account, but usually people don't take NaNs into account when they say things like this.

Reviewer 5: > They talk about "source precision" and "machine precision". So the result of `a + b + c` might be different from what IEEE says since intermediate results might be stored at a higher (or lower?) precision than that of the source type?

This is essentially reference to C's `FLT_EVAL_METHOD` (which is C's solution to the x87 problem). C§5.2.4.2 says:

> The values of floating type yielded by operators subject to the usual arithmetic conversions, including the values yielded by the implicit conversion of operands, and the values of floating constants are evaluated to a format whose range and precision may be greater than required by the type. Such a format is called an evaluation format. In all cases, assignment and cast operators yield values in the format of the type.

> It says that rounding to source precision is done at function calls and returns. So their inliner must preserve these function boundaries then to ensure the value doesn't stay at machine precision?

This is adding function parameters and returns to the list of expressions that force to source precision. It requires the same logic in the optimizer as representing an assignment or a cast expression, so it's not much extra burden to support at all.

> Does this mean they do not even apply commutativity? Commutativity does not produce a bitwise identical result on actual hardware if you take NaNs into account, but usually people don't take NaNs into account when they say things like this.

IEEE 754 has a concept of "value-preserving transformations" and changing the NaN bit pattern is not a "value-preserving transformation." I don't know specifically what MSVC guarantees, but in general, I would not expect NaN payloads to be preserved by optimization.

Author: >  changing the NaN bit pattern is not a "value-preserving transformation." 

>  I would not expect NaN payloads to be preserved by optimization.

Those two statements seem to contradiction each other, is there a negation missing / too much somewhere?

---

But I think we can conclude that MSVC does not document what the permissible bit patterns in NaNs are. So it's probably at least "any qNaN". It's also unclear if they perform any transformations that would make arithmetic return an sNaN, like `x * 1.0 -> x`.

Author: Is `x * 1.0 -> `x` considered a "value-preserving transformation" by IEEE 754? If so, that seems self-contradicting with also saying that arithmetic operations never return a signaling NaN.

Reviewer 5: > Is `x * 1.0 -> x` considered a "value-preserving transformation" by IEEE 754? If so, that seems self-contradicting with also saying that arithmetic operations never return a signaling NaN.

It's explicitly called out as something that is *not* a value-preserving transformation.

Author: Interestingly I just noticed that C18 explicitly calls it out as an allowed transformation. So C seems to be fine with operations returning signaling NaNs, but also signaling NaNs lead to unspecified behavior elsewhere? I guess the idea is that when I pass a signaling NaN to a multiplication, that's already unspecified behavior, and that's why it is okay for the output to violate the usual float semantics rules.

That means the only way to ever see a signaling NaN in a program is to create one with `from_bits` (or equivalent unsafe operations).

Floating-point operations at compile-time follow the same specification. In particular, since operations are non-deterministic, the same operation can lead to different bit-patterns when executed at compile-time (in a `const` context) vs at run-time.

# Motivation
[motivation]: #motivation

We have a plethora of open issues that boil down to "is this sequence of float operations allowed to produce the given result".
This is caused by a combination of surprising effects introduced by LLVM optimizations, bugs in MIR and LLVM optimizations, and bugs (or at least non-conformance) in certain targets. See [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) for a collection of issues.

It's time to stop leaving our users in the dark about what actually is and is not guaranteed.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Primitive operations on floating-point types generally produce results that exactly match IEEE 754-2008:

> if you never use `to_bits`, `from_bits`, `is_sign_negative`, or `is_sign_poitive`, nor any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically according to the IEEE specification.

Reviewer 6: ```suggestion
if you never use `to_bits`, `from_bits`, `is_sign_negative`, or `is_sign_positive`, nor any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically according to the IEEE specification.
```

  (For instance, on wasm, if any input NaN does not have the preferred all-zero   payload or any input NaN is an SNaN, then this set contains all possible payloads; otherwise, it is empty. On SPARC, this set consists of the all-one payload.)

In particular, if all input NaNs are quiet (or if there are no input NaNs), then
the output NaN is definitely quiet. Signaling NaN outputs can only occur if they
are provided as an input value. For example, "fmul SNaN, 1.0" may be simplified
to SNaN rather than QNaN. Similarly, if all input NaNs are preferred (or if
there are no input NaNs) and the target does not have any "extra" NaN payloads,
then the output NaN is guaranteed to be preferred.

The non-deterministic choice happens when the operation is executed; i.e., the result of a NaN-producing floating point operation is a stable bit pattern (looking at these bits multiple times will yield consistent results), but running the same operation twice with the same inputs can produce different results.

> Unless noted otherwise, the same rules also apply to NaNs returned by other library functions (e.g. `min`, `mimimum`, `max`, `maximum`); other aspects of their semantics and which IEEE 754-2008 operation they correspond to are documented with the respective functions.

Reviewer 7: ```suggestion
Unless noted otherwise, the same rules also apply to NaNs returned by other library functions (e.g. `min`, `minimum`, `max`, `maximum`); other aspects of their semantics and which IEEE 754-2008 operation they correspond to are documented with the respective functions.
```

[reference-level-explanation]: #reference-level-explanation

This RFC specifies the behavior of `+`, `-` (unary and binary), `*`, `/`, `%`, `abs`, `copysign`, `mul_add`, `sqrt`, `as`-casts that involve floating-point types, and all comparison operations on floating-point types.
Here, "floating-point types" are `f32` and `f64` and all similar types that might be added in the future such as `f16`, `f128`.
Except for the cases handled below, these operations produce results that exactly match IEEE 754-2008 (with roundTiesToEven and default exception handling, without abruptUnderflow/flush-to-zero).
`%` matches the behavior of `fmod` in C (this operation is not in the IEEE spec).
When a floating-point value is just passed around (i.e., outside the operations above or any library-provided float operation), its representation bits do *not* change.

Exceptions apply when the output of an operation is a NaN, and the operation is not a "bitwise" operation (unary `-`, `abs`, `copysign`).
In that case, we generally follow [the same rules as LLVM](https://llvm.org/docs/LangRef.html#behavior-of-floating-point-nan-values).

To be concrete, we first establish some terminology:

> A floating-point NaN value consists of a sign bit, a quiet/signaling bit, and a payload (which makes up the rest of the mantissa except for the quiet/signaling bit). Rust assumes that the quiet/signaling bit being set to ``1`` indicates a quiet NaN (QNaN), and a value of ``0`` indicates a signaling NaN (SNaN). In the following we will hence just call it the "quiet bit".

Reviewer 8: Nit: mantissa -> significand

Author: Oh that's weird terminology, but Wikipedia agrees...

For the operations listed above, the following rules apply when a NaN value is returned:
the result has a non-deterministic sign; the quiet bit and payload are non-deterministically chosen from the following set of options:

- The quiet bit is set and the payload is all-zero. ("Preferred NaN" case)
- The quiet bit is set and the payload is copied from any input operand that is a NaN.
  ("Quieting NaN propagation" case)
  If the inputs and outputs do not have the same size (i.e., for `as` casts), then
  - If the output is smaller than the input, low-order bits of the payload get dropped.
  - If the output is larger than the input, the payload gets filled up with 0s in the low-order bits.
- The quiet bit and payload are copied from any input operand that is a NaN.
  ("Unchanged NaN propagation" case)

>   If the inputs and outputs do not have the same size (i.e., for `as` casts), the same rules as for "quieting NaN propagation" apply, with one caveat: if the output is smaller than the input, droppig the low-order bits may result in a payload of 0; a payload of 0 is not possible with a signaling NaN (the all-0 mantissa encodes an infinity) so unchanged NaN propagation cannot occur with some inputs.

Reviewer 8: Nit: mantissa -> significand

See the reference section for what exactly is guaranteed.

When a floating-point value is just passed around, its contents (including the bits of a NaN) do *not* change.

When you use a floating-point operation in [`const` context](https://doc.rust-lang.org/reference/const_eval.html#const-context), the same specification applies: NaN bit patterns are non-deterministic.
In particular, the bit pattern produced at compile-time can differ from the bit pattern produced by the same operation at run-time.

Certain targets unfortunately are known to not implement these semantics precisely (see [below](#target-specific-problems)).
The [platform support page](https://doc.rust-lang.org/rustc/platform-support.html) will list those caveats.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

This RFC specifies the behavior of `+`, `-` (unary and binary), `*`, `/`, `%`, `abs`, `copysign`, `mul_add`, `sqrt`, `as`-casts that involve floating-point types, and all comparison operations on floating-point types.
Here, "floating-point types" are `f32` and `f64` and all similar types that might be added in the future such as `f16`, `f128`.

> Except for the cases handled below, these operations produce results that exactly match IEEE 754-2008 (with roundTiesToEven and default exception handling, without abruptUnderflow/flush-to-zero).

Author: I'm actually a bit confused about this point, I copied from somewhere that roundTiesToEven is the default, but `1.8 as i32` returns `1` so actually the default seems to be toward-zero?

Reviewer 9: no, all operations other than conversion to integer default to round-to-nearest-ties-to-even, conversion to integer defaults to truncate (you can think of it as specifically overriding rounding mode for conversion to integer). This is inherited from C afaict, but is standard practice across a wide range of languages. the default for conversion to integer can be overridden using `floor`/`ceil`/`round`/`trunc`/`round_even` (`x.floor() as i32` rounds toward -Infinity)

Author: Oh okay, so this affects *all* operations? But the hardware instructions that we use for float-to-int casts ignore the rounding mode and always round to zero?

Reviewer 9: It affects almost all operations, some use specified rounding modes (e.g. `floor` uses round toward -Infinity).
LLVM intentionally picks round-to-zero instructions for float-to-int casts. e.g. on `x86_64-unknown-linux-gnu`, LLVM picks `cvttss2si` (which rounds to zero) instead of `cvtss2si` (which uses the dynamic rounding mode); and on `i586-unknown-linux-gnu`, LLVM temporarily changes the rounding mode to round toward zero while doing the conversion.

Author: Makes sense, thanks!

This resolves the last open question blocking floating-point operations in `const fn`.
When the RFC is accepted, the `const_fn_floating_point_arithmetic` feature gate and the `const fn to_bits` methods can be stabilized.

### Assumptions about floating-point environment

This RFC is primarily concerned with the guarantee Rust provides to its users.
How exactly those guarantees are achieved is an implementation detail, and not observable when writing pure Rust code.
However, when mixing Rust with inline assembly, those details *do* become observable.
To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.
This is just an instance of the general principle that it is UB for inline assembly to violate any of the invariants that the Rust compiler relies on when implementing Rust semantics on the target hardware.

> Furthermore, observing the floating-point exception state yields entirely unspecified results: Rust floating-point operations may or may not be executed at the place in the code where they were originally written, and the exception state can change even if no floating-point operation exists in the source code.

Author: https://github.com/rust-lang/unsafe-code-guidelines/issues/471 is an objection to this part of the RFC. However it's also not clear whether there are any viable alternatives.

```rust
#![allow(non_snake_case)]
use std::hint::black_box;

const C: f32 = 0.0 / 0.0;

fn main() {
    let c: f32 = 0.0 / black_box(0.0);
    assert_eq!(C.to_bits(), c.to_bits());
}
```

This means that evaluating the same `const fn` on the same arguments can produce different results at compile-time and run-time.
However, note that these functions are already non-deterministic: even evaluating the same function with the same arguments twice at runtime can [and does](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=50b5a549fa1fe259cea5ad138066ccf0) produce different results!

This resolves the last open question blocking floating-point operations in `const fn`.
When the RFC is accepted, the `const_fn_floating_point_arithmetic` feature gate and the `const fn to_bits` methods can be stabilized.

### Assumptions about floating-point environment

This RFC is primarily concerned with the guarantee Rust provides to its users.
How exactly those guarantees are achieved is an implementation detail, and not observable when writing pure Rust code.
However, when mixing Rust with inline assembly, those details *do* become observable.

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Reviewer 8: Possibly, but it would probably have to be fairly specific and would rely on LLVM promising more things. We presently have an example causing a segfault where the only unsafe thing is setting the rounding mode: https://github.com/rust-lang/unsafe-code-guidelines/issues/471#issuecomment-1774261953

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Author: I don't think we can do better than "UB" here. When an operation does not produce the result that the spec says it produces, there's no limit to what happens. People can write `if 1+1 != 2 { unreachable_unchecked() }`, and they can do the same with float operations, so changing the rounding mode can lead to arbitrary misbehavior including UB.

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Reviewer 10: But this is the spec we're writing now, isn't it, so if the spec was weaker they couldn't rely on it? Can we weaken it to say something like "in the absence of code which modifies these flags, floats behave according to IEEE 754-2008" and "if you change any of these flags, optimizations will cause your code to behave nondeterministically"?

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Reviewer 8: What's the difference between "behave nondeterministically" and UB?

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Reviewer 10: Sorry, what I meant was that the _result of floating point operations_ would be nondeterministic. But we can still put bounds around what they would do; e.g. they would always return a value if you haven't set any exception flags.

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Reviewer 8: That's strengthening the spec, not weakening it. In any case, we cannot, as the aforementioned example shows. LLVM will happily optimize away (integer!) array index bounds checks if it can prove that they are never needed under default rounding.

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Author: > we can still put bounds around what they would do

That's exactly the point, we cannot. (a) there is the [example](https://github.com/rust-lang/unsafe-code-guidelines/issues/471#issuecomment-1774261953), and (b) the compile ris allowed to move floating-point operations around, so even if you wrote the FP operation outside the block where the flags are different, it might end up executing inside the block.

I don't know any alternative to declaring this UB.

The other reason is that LLVM is currently architecturally unable to reflect target-dependent behavior in its constant-folder, and the LLVM developers have not shown any interest in providing a guarantee of matching target behavior (except via "strict" floating-point intrinsics which are just not optimized).
It is pragmatically very hard for us to provide guarantees that LLVM does not intend to provide.

Also note that in some cases even the underlying target behavior is non-deterministic, namely on wasm.

### Alternative: operations can never produce a signaling NaN

An operation that returns a signaling NaN violates the basic IEEE guarantee that floating-point operations never produce signaling NaNs. However, LLVM considers it legal to fold `x * 1.0` to `x`, so if `x` is a signaling NaN then the multiplication can have that signaling NaN as the output.
Similarly, `(x: f32) as f64 as f32` may get folded into `x`, so even casts can produce signaling NaNs.

> Given that signaling NaNs are basically useless in Rust since we do not consider the exception status flag to be observable (and hence optimizations can change whether an exception is triggered ot not), there is no point in constraining optimizations just to achieve a guarantee about signaling NaNs.

Reviewer 10: Ah, I just saw the reference to the last point at the end of the draft. And given https://github.com/rust-lang/rfcs/pull/3514#issuecomment-1778335626 I think this is resolved.

### Alternative: `const` tracks NaN values, fails when their bits matter during compile-time

`const` could in principle track NaNs symbolically, similar to how it tracks pointers, and delay choosing NaN payload bits until codegen.
Const-evaluation would then abort if the bits of a NaN are observed (eg. if `to_bits`is called during const-evaluation).
This would keep `const C = 0.0/0.0;` working, but requires `const fn is_nan` to be an intrinsic.
However, it would require massive amounts of work in the compile-time interpreter, comparable in complexity to all the work that is already required to support symbolic pointers (and the RFC author doubts that there will be a lot of opportunity for those two kinds of symbolic state to share infrastructure).
That effort should only be invested if there is a significant payoff.
The RFC author considers the downsides of unspecified NaN bit patterns being observable in const-evaluation to be minimal, and hence the payoff of this alternative to be low.

# Prior art
[prior-art]: #prior-art

> C and C++ generally avoid saying anything about signaling NaNs, and also don't talk about how NaNs are chosen. I can only assume this means they are chosen non-deterministically; that is certainly the interpretation chosen by GCC and LLVM.

Reviewer 5: > C and C++ generally avoid saying anything about signaling NaNs

This isn't quite accurate. C17 and prior has explicit wording

> This specification does not define the behavior of signaling NaNs

which makes it undefined behavior (although this tends to be in the sense of "extension point for implementers" rather than "assume it can't happen"). C23 changes this wording to say that

> Where specification of signaling NaNs is not provided, the behavior of signaling NaNs is implementation-defined (either treated as an IEC 60559 quiet NaN or treated as an IEC 60559 signaling NaN).

LLVM presently is closer to the former definition.

Author: > This specification does not define the behavior of signaling NaNs

That's what I meant when I said they avoid saying things about it. It's not clear to me that this really means that creating a signaling NaN via bit reinterpretation is UB.

I didn't realize C23 is updating that, thanks for pointing that out! I will update the RFC.

Author: (update done)

> GCC [says](https://gcc.gnu.org/wiki/FloatingPointMath) "Without any explicit options, GCC assumes round to nearest or even and does not care about signalling NaNs". It is unclear whether "does not care" also means "guarantees to never produce by itself", i.e. whether `0.0 / 0.0` is ever allowed to evaluate to a signaling NaN or not. If it *is* allowed to evaluate to a signaling NaN, that is probably a violation of the C standard, which guarantees that `pow(1, 0.0/0.0)` returns `1` -- but in practice, `pow(1, sNaN)` returns a NaN.

Reviewer 5: I don't have firm records of the history of CFP deliberation on the wording of C §F.2.1. IEEE 754 makes it clear that sNaNs are not producible by regular operations under normal situations, and the footnote attached to the statements quoted above in §F.2.1 also makes it clear that nonsupport for sNaN isn't an issue:

> Since NaN s created by IEC 60559 arithmetic operations are always quiet, quiet NaNs (along with infinities) are sufficient for closure of the arithmetic.

`0.0 / 0.0` can only produce a qNaN, even in don't-care-about-sNaN mode.

Author: Yeah, under IEEE 754 there certainly can't be any sNaN produced by arithmetic operations.

But under optimizing compilers, sNaN *can* be produced by arithmetic operations, so looking at the IEEE spec won't help us here. We have to look at the compiler docs and the specs of the language they implement. For LLVM we now have very clear docs telling us when arithmetic can produce an sNaN (in particular, `0.0 / 0.0` can *not* produce an sNaN). I am not aware of any such documentation for GCC.

So, if we say that any operation can arbitrairly produce signaling NaNs, then it becomes impossible to rely on `pow`'s property that "`pow(+1, exponent)` returns `1` for any `exponent`".
Since the quiet/signaling distinction matters, we provide a guarantee which ensures that programs will almost always only ever deal with quiet NaNs (the only safe way to get a signaling NaN is to use `from_bits`).

### Alternative: no strict semantic guarantees

Providing strict IEEE 754-2008 guarantees precludes many transformations, such as turning `a*b + c` into FMA operations (since the result can change due to different rounding).
We could provide weaker guarantees to allow such transformations.
(That would also make using the higher-precision x87 instructions conforming with the spec.)
However, there is no way to bound the effect of such transformations on program behavior: they lead to different outcome observable via `to_bits` and even via pure float operations, which can balloon into arbitrary changes in behavior.

> There is also no "monotonicity of precision": while e.g. an FMA instead of mul-then-add will lead to a higher-precision result for this particular operation, this can in principle lead to a lower-precision result in later computations.

Reviewer 11: The classic example here is evaluating `x*x - x*x` as `x.mul_add(x, -x*x)` producing nonzero. This is a bit awkward because `fma` is slow on x86_64 generic so you don't want to use it unless you need it, but fast when target features are enabled (on host or device). One could exclude [`mul_add`](https://doc.rust-lang.org/std/primitive.f64.html#method.mul_add) from `const fn`, but outside of specific cases where symmetry in such expressions are necessary, scientific computing users will usually prefer `mul_add` semantics (both for accuracy and performance). This is especially relevant to GPU targets where FMA is heavily used. It would be a shame to eternally forbid using `mul_add` in `const fn` (thereby degrading its accuracy everywhere and reducing performance when called at run-time). I'm not sure what the right semantic is here, but it's an important niche, I'm concerned that the present semantic may be too strict, and I'm not sure a "fast float type" covers bases in a useful way. (Among other things, trait methods cannot be `const`.)

Author: I don't understand which point you are trying to make here. Nobody is suggesting to eternally forbid `mul_add` in `const fn`. This paragraph you are replaying to is about optimizations that automatically introducing FMA into code that didn't use FMA -- the RFC says that that is not permitted.

Reviewer 11: Thanks for clarifying. Lack of support for [`AllowContract`](https://llvm.org/doxygen/classllvm_1_1FastMathFlags.html) is quite painful for people working in scientific computing, and is responsible for a sizeable fraction of the case studies "I ported {physics kernel} from C to Rust and it's much slower" (the rest having easier resolutions, like checked indexing).

The only way to get FMAs with Rust (apart from direct use of `mul_add`, which has bad performance consequences on some architectures and is awkward to program with) is nightly-only (`core_intrinsics` to access the likes of `fadd_fast` and `fmul_fast`, which will always be unsafe sledgehammers). There have been suggestions that new intrinsics like `fadd_with_flags(a, b, flags)` could possibly be stabilized so long as flags like `NoInfs` were unsafe. As I understand it, the proposed semantic here would prevent that from ever being possible (even `AllowContract`, which is the default behavior of many C compilers). Moreover, the strategy of [fast-floats](https://crates.io/crates/fast-floats) would effectively require that all perf-sensitive computation become unsafe (it's hard enough to prove that values are always remain finite, but observable differences between FMA and non-FMA optimization paths are everywhere). This would be a shame since safe Rust can provide immense value to that community.

A different way to ask: is there a way we could leave the door open for a future proposal to bring `AllowContract` semantics in safe Rust?

Author: This RFC does not close the door to fast-math operations (Cc https://github.com/rust-lang/rust/issues/21690). It just says they must be opt-in. Otherwise it's entirely orthogonal.

Reviewer 9: > The only way to get FMAs with Rust (apart from direct use of `mul_add`, which has bad performance consequences on some architectures

I've wanted Rust to have a function that translates to the LLVM [`llvm.fmuladd.*` intrinsic](), which is always either a FMA or separate mul and add, no other options.

Reviewer 11: > I see no reason why they must be unsafe

Certainly `fadd_fast`/`fmul_fast` must remain unsafe because `NoNaNs` and `NoInfs` [produce poison](https://llvm.org/docs/LangRef.html#fast-math-flags). The other flags enable target- and optimization-dependent results, but this discussion clarifies that we agree that is okay in safe code.

I was put off by this RFC reading as though NaN-nonuniqueness was a limited carve-out of the convention that runtime and const contexts produce the same results. It looks like [this suggested RFC](https://github.com/rust-lang/rust/issues/77745#issuecomment-1207286013) was never written, but would clarify what are widely varying opinions in the community. I think default semantics leaning toward deterministic is good so long as we don't inhibit an easy/safe/stable way to opt in.

Thanks @programmerjake, I think you meant to link [this](https://llvm.org/docs/LangRef.html#llvm-fmuladd-intrinsic). The question (for a different thread) is how to expose it.

Author: > Certainly fadd_fast/fmul_fast must remain unsafe because NoNaNs and NoInfs [produce poison](https://llvm.org/docs/LangRef.html#fast-math-flags).

Oh sure. But the basic idea of "rustc may optimize assuming the laws of the regular rational numbers" doesn't need unsafe I think. It might just be that the way LLVM exposes it needs unsafe.

> I was put off by this RFC reading as though NaN-nonuniqueness was a limited carve-out of the convention that runtime and const contexts produce the same results. 

We do plan to allow const fn to arbitrarily differ from runtime behavior, but that's a different RFC: https://github.com/rust-lang/rfcs/pull/3352.

This RFC just says that const fn can be non-deterministic when executed at runtime, but the compile-time behavior is still in the set of possible rutnime behaviors. (But "possible" must be interpreted in a spec sense here: the spec allows that behavior to occur. The behavior might never actually happen with current rustc.)

Given all these caveats, it seems preerrable to have explicit opt-in for such semantics-alterating transformations, e.g. via "fast" floating point operation intrinsics that don't provide strict IEEE semantics (and a type that uses those intrinsics for its operations).

## Alternative options for `const` semantics

### Alternative: `const` must be deterministic, and match runtime behavior

This RFC proposes that we accept, for the first time, that a `const fn` can behave non-deterministically at run-time, and produce target/version/flag-specific results at compile-time. It can also produce different results when called at compile-time and at run-time.

> [Another proposed RFC](https://github.com/rust-lang/rfcs/pull/3352) gathers some general arguments for why we should allow such behavior in a `const fn`.

Reviewer 12: Does this mean that if this RFC is merged, the other one vam be closed as this implies the other one? I think it makes more sense or accept the other one first instead of doing this semantic change here as a sidenote in this RFC.

Author: For the other RFC we were told we need a concrete instance  of some behavior we want to add to const-eval, we shouldn't just discuss and decide such things in the abstract. So I don't think we can land the other one first.

We could split this up, make this RFC only about runtime float semantics and then once it lands make a second RFC about compiletime float semantics. But I thought it makes more sense to discuss float semantics in general. I don't view this as just a sidenote in this RFC.

Of course, the compile-time interpreter is still deterministic. It is entirely possible to implement a non-deterministic language on a deterministic machine, by simply making some fixed choices. However, we will not specify a particular choice, and we will not guarantee it to remain the same in the future.

# Motivation
[motivation]: #motivation

We have a plethora of open issues that boil down to "is this sequence of float operations allowed to produce the given result".
This is caused by a combination of surprising effects introduced by LLVM optimizations, bugs in MIR and LLVM optimizations, and bugs (or at least non-conformance) in certain targets. See [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) for a collection of issues.

It's time to stop leaving our users in the dark about what actually is and is not guaranteed.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Primitive operations on floating-point types generally produce results that exactly match IEEE 754-2008:

> if you never use `to_bits`, `copysign`, `is_sign_negative`, or `is_sign_positive` on a NaN, and don't construct a NaN using `from_bits`, nor use any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically and according to the IEEE specification.

Author: That's just a misconception. These operations make prior non-determinism observable.

> Also, the standard does consider changing the sign and payload bits of a quiet NaN to preserve the literal meaning of the source code, meaning that transformations that do so are allowed by default.

No that's not true. The RFC explicitly states that just copying a float around will not change its representation bits.

> if you never use `to_bits`, `copysign`, `is_sign_negative`, or `is_sign_positive` on a NaN, and don't construct a NaN using `from_bits`, nor use any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically and according to the IEEE specification.

Reviewer 13: Yes, I fully agree that Rust code that just copies floats around should not see those bits changing, and any divergence from that is a bug in the implementation (as with the case of returning floats via the x87 FPU). The point was that only floating point operations other than those sign-operations may create a nondeterministic NaN, and thus lead to nondeterminism.

I see I worded the second point unclearly. When I said "changing", I did not mean inserting an operation that changes the bits, but transforming a computation that produces one NaN-payload into a computation that produces a different NaN-payload.

The IEEE754 standard says that language standards should require implementations to preserve the "literal meaning" of the source code (by default), which includes numeric results and the set of status flags raised, but not the sign or payload of any *quiet* NaN. This is what allows things like "the NaN-creating instruction stashes some instruction pointer bits into the payload", since that would obviously make the payloads observed at runtime very nondeterministic.

> if you never use `to_bits`, `copysign`, `is_sign_negative`, or `is_sign_positive` on a NaN, and don't construct a NaN using `from_bits`, nor use any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically and according to the IEEE specification.

Author: Rust chooses to ignore the part about preserving the set of status flags raised. Status flags are not considered observable. And Rust also largely ignores the quiet/signalling distinction. This is in line with common practice in C and C++ compilers.

I have reworded this sentence a bit. I don't know if you are asking for anything else to be changed about the RFC.

> if you never use `to_bits`, `copysign`, `is_sign_negative`, or `is_sign_positive` on a NaN, and don't construct a NaN using `from_bits`, nor use any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically and according to the IEEE specification.

Reviewer 13: I wanted it to instead say something along the lines of "If your code does not depend on the sign and payload of any quiet NaN, then it cannot observe NaN-related nondeterminism.", although that feels a bit tautological. (Feel free to drop and ignore the "quiet" since we don't properly support sNaN. (If we did, it would matter.))

The problem with claiming to list all sources of something is that it's easy to miss some. `total_cmp` is another safe method that can distinguish NaNs, and then there's the separate issue of nondeterministic numeric values from any of the functions that are allowed some leeway in precision.

Exceptions apply when the output of an operation is a NaN, and the operation is not a "bitwise" operation (unary `-`, `abs`, `copysign`).
In that case, we generally follow [the same rules as LLVM](https://llvm.org/docs/LangRef.html#behavior-of-floating-point-nan-values).

To be concrete, we first establish some terminology:
A floating-point NaN value consists of a sign bit, a quiet/signaling bit, and a payload (which makes up the rest of the significand (i.e., the mantissa) except for the quiet/signaling bit). Rust assumes that the quiet/signaling bit being set to ``1`` indicates a quiet NaN (QNaN), and a value of ``0`` indicates a signaling NaN (SNaN). In the following we will hence just call it the "quiet bit".

For the operations listed above, the following rules apply when a NaN value is returned:
the result has a non-deterministic sign; the quiet bit and payload are non-deterministically chosen from the following set of options:

- The quiet bit is set and the payload is all-zero. ("Preferred NaN" case)
- The quiet bit is set and the payload is copied from any input operand that is a NaN.
  ("Quieting NaN propagation" case) <br>

>   If the inputs and outputs do not have the same size (i.e., for `as` casts), then

Reviewer 9: I think it should be clarified here that `size` refers to the size of the payload, since otherwise `f16` and `bfloat16` are both 16-bits, but conversions still need to add/drop payload bits.

> - The quiet bit is set and the payload is all-zero. ("Preferred NaN" case)
> - The quiet bit is set and the payload is copied from any input operand that is a NaN.
>   ("Quieting NaN propagation" case) <br>
>   If the inputs and outputs do not have the same size (i.e., for `as` casts), then
>   - If the output is smaller than the input, low-order bits of the payload get dropped.
>   - If the output is larger than the input, the payload gets filled up with 0s in the low-order bits.
> - The quiet bit and payload are copied from any input operand that is a NaN.
>   ("Unchanged NaN propagation" case) <br>
>   If the inputs and outputs do not have the same size (i.e., for `as` casts), the same rules as for "quieting NaN propagation" apply, with one caveat: if the output is smaller than the input, droppig the low-order bits may result in a payload of 0; a payload of 0 is not possible with a signaling NaN (the all-0 significand encodes an infinity) so unchanged NaN propagation cannot occur with some inputs.
> - The quiet bit is set and the payload is picked from a target-specific set of
>   "extra" possible NaN payloads. The set can depend on the input operand values.
>   This set is empty on x86, ARM, and RISC-V (32bit and 64bit), but can be non-empty on other architectures. Targets where this set is non-empty should document this in a suitable location, e.g. their platform support page.
>   (For instance, on wasm, if any input NaN does not have the preferred all-zero   payload or any input NaN is an SNaN, then this set contains all possible payloads; otherwise, it is empty. On SPARC, this set consists of the all-one payload.)

Reviewer 13: This section would be easier to understand if it separated the cases under "the quiet bit is set" and "the quiet bit is copied"
```suggestion
- The quiet bit is set, and
  - the payload is all-zero. ("Preferred NaN" case)
  - the payload is copied from any input operand that is a NaN. ("Quieting NaN propagation" case)  
  - the payload is picked from a target-specific set of "extra" possible NaN payloads. The set can depend on the input operand values. This set is empty on x86, ARM, and RISC-V (32bit and 64bit), but can be non-empty on other architectures. Targets where this set is non-empty should document this in a suitable location, e.g. their platform support page. (For instance, on wasm, if any input NaN does not have the preferred all-zero payload or any input NaN is an SNaN, then this set contains all possible payloads; otherwise, it is empty. On SPARC, this set consists of the all-one payload.)
- The quiet bit and payload are copied from any input operand that is a NaN. ("Unchanged NaN propagation" case)

When the payload is copied from an input with a different size (i.e., for `as` casts), then
- If the output is larger than the input, the payload gets filled up with zeros in the low-order bits.
- If the output is smaller than the input, low-order bits of the payload get dropped.

If the quiet bit and payload would be copied from a larger payload, dropping the low-order bits could result in a payload of 0; a payload of 0 is not possible with a signaling NaN (the all-0 significand encodes an infinity) so unchanged NaN propagation cannot occur with some inputs.
```

Author: I don't think this is better. It puts undue emphasis on the quiet bit.

Reviewer 13: Okay, that's fair. Adding that emphasis was part of my intention, because not quieting an sNaN is the only case which diverges from the IEEE standard. In particular, the RFC seems to allows propagation of sNaN through "x * -1.0" without flipping the sign bit, whereas the sign of an sNaN should behave deterministically in all operations that don't quiet it.

Author: Yes we do deviate from IEEE deliberately here. But I assume few readers to be already familiar with what exactly IEEE says about SNaN so I didn't want to structure this as a diff to IEEE.

If x is an SNaN, then `x * -1.0` has arbitrary sign and is allowed to be
- the preferred QNaN
- a QNaN with the same payload as `x`
- an SNaN with the same payload as `x`

We also guarantee *more* than IEEE guarantees when it comes to the payload bits, so in the end it's incomparable.

Author: I'll leave this thread unresolved to see how others feel about presenting this as a flat list of 4 cases, vs structuring this around the quiet bit.

Reviewer 13: Yes, and according to my understanding, when `x`  is an SNaN, IEEE requires
 - `x * -1.0` -> QNaN with unspecified sign and payload
 - `-x` -> SNaN with the payload of `x` and the opposite sign
 
(the latter is also required in this RFC)

 That is why I previously suggested the weaker (than IEEE) guarantee (see below) that would allow changing `x * -1.0` into `-x` (which LLVM currently does), while still preserving the property that the SNaN behaves like a signed value with unknown magnitude, until some operation depends on the magnitude and quiets it. I believe that would allow all practical optimizations while keeping to the spirit of the standard.

> If some part of a Rust floating point computation is equivalent to an operation
purely on the sign bit of a numeric input, then that same sign operation may be
used to perform the computation even when the input may be NaN.

Reviewer 9: iirc wasm rules deterministic sign out, since basically every arithmetic operation is specified to return non-deterministic sign when returning a NaN

Reviewer 13: Yeah, seems reasonable as a future option, and it is a subset from the current rfc.

With regards to LLVM's current optimizations; Alive2 does seem to implement the behavior as specified in the RFC (sign unspecified even in the snan-propagation case). I haven't managed to construct an example where LLVM would actually optimize like that, but probably it can.

>I think you want to make the sign deterministic when there is a signalling NaN input, but what is the deterministic spec for that sign? Given an arbitrary binary float operation, what should the sign of the returned NaN be?

To be clear, this only concerns the sign of any *signalling* NaNs.

(This actually got more optimization-friendly while formulating the response...)
Let `f(s,m)` be an operation with input and output in (sign,magnitude)-form (i.e. the sign bit `s` and the rest of the bits `m` separately). If a substitution `g(s,m)` is correct for all non-signalling inputs, and does not depend on the bit-representation of `m` (via `total_cmp`, `to_bits`,`is_nan`, etc.), the substitution could be made even if the input could be a signalling NaN. (For the record, this does not conform to the standard, but it's the closest that I've come up with while still allowing optimizations LLVM actually performs.)

So, why would this be reasonable? Because either the optimized code passes the SNaN to a different float-op that quiets it, or the SNaN magnitude gets returned as is, or the output doesn't depend on it at all. Some allowed transforms include:
    - `x * -1.0` -> `-x`
    - `x * 2.0` -> `x + x`
    - `powf(1.0, x)` -> `1.0`
    - `powf(x, 0.0)` -> `1.0`
    - `hypot(inf, x)` -> `inf`
    - `hypot(0, x)` -> `abs(x)`
(The first four of those are currently performed by LLVM)

> iirc wasm rules deterministic sign out, since basically every arithmetic operation is specified to return non-deterministic sign when returning a NaN

They do however [specify](https://webassembly.github.io/spec/core/exec/numerics.html#nan-propagation) that any returned NaN is "arithmetic", which is their definition for "the payload has its most significant bit set", which just so happens to be the quiet bit (but Wasm doesn't make the quiet/signalling distinction). So at least in this regard their rules follow the standard.

Author: @quaternic that characterizes the optimizations, but it's not what we'd need to adjust the semantics. We need a way to say what are all the possible behaviors of a Rust program after applying arbitrarily many of your and other optimizations.

Put differently: currently Miri carries a [pretty direct implementation](https://github.com/rust-lang/miri/blob/1bfe76b6bd4f533591897d32caf17f074b55b8a7/src/operator.rs#L97-L119) of the wording from this RFC, making sure we explore all possible options when a test is repeated sufficiently often. How would that change if we wanted to get your guarantee?

However, this is also getting somewhat off-topic for this RFC since we're not going to have this guarantee until someone convinces LLVM to provide it.  I also don't see the motivation for this change. Why is it useful to have a deterministic sign if the output is a signalling NaN? The quiet/signalling NaN distinction is largely useless anyway  given existing compiler transformations, and the sign of a NaN is still non-deterministic if the NaN is quiet (which is always a possibility), so... it seems like a completely pointless complication of our spec to me to try and make SNaN signs deterministic.

Reviewer 13: Consider this program, which itself uses the default error handling, and calls a library that may utilize SNaNs as reserved operands
```rust
let mut data = somelib::init();
for x in &mut data {
    // negate
    *x = -(*x);
    // alternative: multiply by -1.0
    // *x = *x * -1.0;
}
somelib::eval(data);
```
As written (using `-x`), this program works correctly with this RFC: If there are signalling NaNs in the data initialized by the library, they are negated the same as any numeric values.

If the program instead uses the alternative expression `x * -1.0`, and the IEEE standard was followed, the library would receive quiet NaNs, and report an error (or return a quiet NaN to denote one).

However, the non-conforming specification proposed in the RFC allows the alternative program to instead change the signs in the data arbitrarily without quieting the SNaNs. The library would then *silently* produce incorrect results without either party realizing that an error should have occurred.

If allowing only the limited optimization scheme I described, the program would at least be guaranteed to behave as if it did a nondeterministic choice between  `-x` and `x * -1.0`, giving correct results or reporting errors, respectively, but never silently incorrect results.

Author: My claim, as above, is that the entire distinction of signalling and quiet NaNs is entirely pointless. IEEE had some plans here, and there is an alternative universe in which those planned panned out and they are useful. But in this universe, with today's compilers, they are not. You can't meaningfully use them as reserved values as one can freely create them with `from_bits`. You can't check whether they flow into arithmetic since you can't observe the status bits. You can't check if there was an arithmetic operation since SNaNs might or might not be quieted. So it's not worth trying to support them in any meaningful way, other than making sure we don't generate them "out of thin air".

Anyone who begs to differ is free to write up an RFC detailing documentation and specification; I don't consider that a good use of my time.

Reviewer 13: I could continue with more opinions on your other points, but it indeed isn't very relevant here. We certainly agree that this RFC can't make more guarantees on this unless LLVM does.

```rust
#![allow(non_snake_case)]
use std::hint::black_box;

const C: f32 = 0.0 / 0.0;

fn main() {
    let c: f32 = 0.0 / black_box(0.0);
    assert_eq!(C.to_bits(), c.to_bits());
}
```

This means that evaluating the same `const fn` on the same arguments can produce different results at compile-time and run-time.
However, note that these functions are already non-deterministic: even evaluating the same function with the same arguments twice at runtime can [and does](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=50b5a549fa1fe259cea5ad138066ccf0) produce different results!

This resolves the last open question blocking floating-point operations in `const fn`.
When the RFC is accepted, the `const_fn_floating_point_arithmetic` feature gate and the `const fn to_bits` methods can be stabilized.

### Assumptions about floating-point environment

This RFC is primarily concerned with the guarantee Rust provides to its users.
How exactly those guarantees are achieved is an implementation detail, and not observable when writing pure Rust code.
However, when mixing Rust with inline assembly, those details *do* become observable.

> To ensure that Rust can provide the above guarantees to user code, it is UB for inline assembly to alter the behavior of floating-point operations in any way: when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered.

Reviewer 9: I think this should be clarified to state that it's fine to change the floating-point environment inside the inline asm, but that it needs to be restored before reentering any Rust/C/etc. code. Also, we'll probably want to state that this is referring specifically to fp control flags, changing the fp status bits is fine (maybe only if they're properly marked as clobbered or something).

Author: > I think this should be clarified to state that it's fine to change the floating-point environment inside the inline asm, but that it needs to be restored before reentering any Rust/C/etc. code.

It already clarifies:

"when leaving the inline assembly block, the floating-point environment must be in exactly the same state as when the inline assembly block was entered."

I don't think there is any need for further clarification here.

# Summary
[summary]: #summary

This RFC proposes a specification for how our floating point operations are expected to behave.
The current implementation in rustc already matches the specification, so after accepting the RFC no compiler changes are required.
(However we might be able to stabilize some `const fn` features, see below for details.)

Rust's floating point operations follow IEEE 754-2008 -- with some caveats around operations producing NaNs: IEEE makes almost no guarantees about the sign and payload bits of the NaN; however, actual hardware does not pick those bits completely arbitrarily, and Rust will expose some of those hardware-provided guarantees to programmers.
On the flip side, NaN generation is non-deterministic: running the same operation on the same inputs several times can produce different results.
And there is a caveat: while IEEE specifies that float operations can never output a signaling NaN, Rust float operations *can* produce signaling NaNs, *but only if* an input is signaling.
That means the only way to ever see a signaling NaN in a program is to create one with `from_bits` (or equivalent unsafe operations).

> Floating-point operations at compile-time follow the same specification. In particular, since operations are non-deterministic, the same operation can lead to different bit-patterns when executed at compile-time (in a `const` context) vs at run-time.

Reviewer 14: ```suggestion
Floating-point operations at compile-time follow the same specification. In particular, since operations involving NaN bit patterns are non-deterministic, the same operation can lead to different NaN bit patterns when executed at compile-time (in a `const` context) vs at run-time.
```

To make it more obvious this RFC is only documenting non-determinism for NaNs, not for floating point in general.

In particular, if all input NaNs are quiet (or if there are no input NaNs), then
the output NaN is definitely quiet. Signaling NaN outputs can only occur if they
are provided as an input value. For example, "fmul SNaN, 1.0" may be simplified
to SNaN rather than QNaN. Similarly, if all input NaNs are preferred (or if
there are no input NaNs) and the target does not have any "extra" NaN payloads,
then the output NaN is guaranteed to be preferred.

The non-deterministic choice happens when the operation is executed; i.e., the result of a NaN-producing floating point operation is a stable bit pattern (looking at these bits multiple times will yield consistent results), but running the same operation twice with the same inputs can produce different results.

Unless noted otherwise, the same rules also apply to NaNs returned by other library functions (e.g. `min`, `minimum`, `max`, `maximum`); other aspects of their semantics and which IEEE 754-2008 operation they correspond to are documented with the respective functions.

> ### `const` semantics

Author: Note to self: clarify that we guarantee very little about how this is made "deterministic".

Reviewer 14: Capturing a point from the design meeting discussion:

You can still reasonably decide to evaluate a const fn *fewer* times than the program asks for, since you already have to accept that the result is non-deterministic, so giving the same result as the previous call is one valid outcome. You just can't evaluate it *more* times than the program asks for, because the evaluations might give different results.

# Drawbacks
[drawbacks]: #drawbacks

- This RFC is too restrictive for some targets and too vague for some users: it is too vague since NaN signs are still left completely non-deterministic, which is not actually the case on hardware.
  It is also too restrictive as shown by the following subsection.
- It [looks](https://reviews.freebsd.org/D33599) like some targets, such FreeBSD, leave the floating-point environment in signal handlers in an unspecified state.
  This is something that [Linux specifically avoided](https://yarchive.net/comp/linux/fp_state_save.html) since it breaks e.g. the glibc `memcpy` implementation (which uses SSE registers if available).
  Rust (just like C without `#pragma STD FENV_ACCESS`) cannot currently be used to write signal handlers on such targets.
  There is little we can do here with the current state of LLVM; and even once LLVM provides the necessary features, these signal handlers will need annotations in the code that tell the compiler about the non-default floating point state.
  Those targets chose to use a semantics that is hard to support well in a highly optimized language, and there's not much we can do to paper over such target quirks.

>   We can only hope that eventually those targets will decide to provide a reliable floating-point environment.

Author: I should add that this can be mostly worked around with inline assembly that sets the floating-point environment to the default state Rust accepts. While strictly speaking that would have to happen before any Rust function gets called, practically speaking if this is the first thing the signal handler does, it is very unlikely that the compiler will put a floating-point operation between the function start and the inline assembly block.

In Rust itself, questions around float semantics have been discussed for a long time.
[This issue](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) collects a lot of that discussion, which culminated in this RFC.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Should we be concerned that LLVM does [not actually document](https://github.com/llvm/llvm-project/issues/60942) that it uses IEEE float semantics?
  It does assume IEEE semantics in its own optimization passes.
- Are there any other targets with floating-point trouble?
- What exactly is the set of "extra" NaNs for all remaining targets?
- To what extend does this specification apply to platform intrinsics?
  On the one hand, it seems reasonable to expect platform intrinsics to have the behavior of the platform instructions.
  On the other hand, we implement some platform intrinsics with the portable LLVM `simd` intrinsics, and those are subject to the NaN-non-determinism described above.
  So the current de-facto semantics of at least some platform intrinsics is that they do *not* match what the platform does.

# Future possibilities

> [future-possibilities]: #future-possibilities

Author: We could have a lint that triggers when a const op generates a float.

Reviewer 14: @RalfJung I think you meant "generates a NaN" here?

Author: Ah, yes.^^ Fixed.

See the reference section for what exactly is guaranteed.

When a floating-point value is just passed around, its contents (including the bits of a NaN) do *not* change.

When you use a floating-point operation in [`const` context](https://doc.rust-lang.org/reference/const_eval.html#const-context), the same specification applies: NaN bit patterns are non-deterministic.
In particular, the bit pattern produced at compile-time can differ from the bit pattern produced by the same operation at run-time.

Certain targets unfortunately are known to not implement these semantics precisely (see [below](#target-specific-problems)).
The [platform support page](https://doc.rust-lang.org/rustc/platform-support.html) will list those caveats.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

This RFC specifies the behavior of `+`, `-` (unary and binary), `*`, `/`, `%`, `abs`, `copysign`, `mul_add`, `sqrt`, `as`-casts that involve floating-point types, and all comparison operations on floating-point types.
Here, "floating-point types" are `f32` and `f64` and all similar types that might be added in the future such as `f16`, `f128`.

> Except for the cases handled below, these operations produce results that exactly match IEEE 754-2008 (with roundTiesToEven [except for float-to-int casts, which round towards zero] and default exception handling without traps, without abruptUnderflow/flush-to-zero).

Reviewer 15: ```suggestion
Except for the cases handled below, these operations produce results that exactly match IEEE 754-2008 (with roundTiesToEven [except for float-to-int casts, which round towards zero](...) and default exception handling without traps, without abruptUnderflow/flush-to-zero).
```

The link seems to miss a URL.

Author: No, these are just nested parentheses.

Reviewer 15: Oh, I see. Thanks for the clarification.
